include/rocksdb/multi_scan_iterator.h

```diff
@@ -0,0 +1,264 @@
+#pragma once
+
+#include "rocksdb/iterator.h"
+#include "rocksdb/options.h"
+
+namespace ROCKSDB_NAMESPACE {
+
+#if 0
+// An iterator that returns results from multiple scan ranges. The ranges are
+// expected to be in increasing sorted order. The application on top of RocksDB
+// would use this as follows -
+//
+//  std::vector<ScanOptions> scans{{.start = Slice("foo")},
+//                              {.start = Slice("bar")}};
+//  std::unique_ptr<MultiScanIterator> iter.reset(
+//                                      db->NewMultiScanIterator());
+//  for (auto& scan : scans) {
+//    PinnableSlice val;
+//    while (iter-status().ok() && !iter->empty()) {
+//      Slice key;
+//      std::string value;
+//      iter->Dequeue(key, val);
+//      if (val.IsPinned()) {
+//        val_str = val.ToString();
+//      } else {
+//        val_str = std::move(*val.GetSelf());
+//      }
+//      val.Reset();
+//      // Do something with key and val_str
+//    }
+//    if (!iter->status().ok()) {
+//      break;
+//    }
+//    assert(iter->empty());
+//    iter->SeekNext();
+//  }
+//  assert(!iter->status().ok() || iter->empty());
+//
+class MultiScanIterator {
+ public:
+  MultiScanIterator(const std::vector<ScanOptions>& scans,
+                    std::unique_ptr<Iterator>&& iter)
+      : scan_opts_(scans), idx_(0), iter_(std::move(iter)) {
+    // Position the iterator for the first scan
+    NextScan();
+  }
+
+  void Dequeue(Slice& key, PinnableSlice& value) {
+    key = iter_->key();
+    value.PinSelf(iter_->value());
+    iter_->Next();
+    if (!iter_->Valid()) {
+      empty_ = true;
+    }
+    status_ = iter_->status();
+  }
+
+  void SeekNext() {
+    idx_++;
+    if (idx_ < scan_opts_.size()) {
+      NextScan();
+    } else {
+      empty_ = true;
+    }
+  }
+
+  bool empty() { return empty_; }
+
+  Status status() { return status_; }
+
+ private:
+  const std::vector<ScanOptions>& scan_opts_;
+  size_t idx_;
+  std::unique_ptr<Iterator> iter_;
+  bool empty_;
+  Status status_;
+
+  void NextScan() {
+    iter_->Seek(scan_opts_[idx_].start);
+    empty_ = !iter_->Valid();
+    status_ = iter_->status();
+  }
+};
+#endif
+
+// EXPERIMENTAL
+//
+// An iterator that returns results from multiple scan ranges. The ranges are
+// expected to be in increasing sorted order. The application on top of RocksDB
+// would use this as follows -
+//
+//  std::vector<ScanOptions> scans{{.start = Slice("bar")},
+//                              {.start = Slice("foo")}};
+//  std::unique_ptr<MultiScanIterator> iter.reset(
+//                                      db->NewMultiScanIterator());
+//  try {
+//    for (auto scan : *iter) {
+//      for (auto it : scan) {
+//        // Do something with key - it.first
+//        // Do something with value - it.second
+//      }
+//    }
+//  } catch (Status s) {
+//  }
+class MultiScanIterator {
```

Comment:
I found it difficult to read the giant nesting of classes below to deduce the relationship between MultiScanIterator, ScanIterator, Scan, and SingleIterator. Could you add some diagram or hierarchy or elaboration to the code example to make that clear?

Also should `MultiScanIterator` just be called `MultiScan` since it doesn't behave like an STL or RocksDB iterator?